### ISSUE TITLE
Before/After event for library file functions

### ISSUE THREAD

**Issue by punitkumar** [opencart/opencart#11793]
**What version of OpenCart are you reporting this for?**
Opencart4x

**Describe the bug**
Alike before/after events for controller/model/view/language/ 
I am not able to make events that will trigger upon library function call, for example: 
library cart.php function getProducts()

Many modules that are dependent upon return value of cart library functions not able to converted into Opencart4x .

There is a approach to extend cart library class then override cart object in registry.

https://github.com/opencart/opencart/discussions/11624

However, the above approach does not works when two modules follow same approach to accomplish the module functionality. 

There is another suggestion like make event for all controller, model files in which library cart function get called, but this approach is not feasible because we never know if any third party module is calling the cart functions. This approach does not work if core controller file has redirect condition based on cart function return value. 

**Expected behavior**
There must a way to register an event for library files or limited files, in order to all extensions to function smoothly.

**Screenshots / Screen recordings**
Have look at following video that explain issue when implementing library file class extend approach
https://www.loom.com/share/7e0c628ed42b45b2beca93999fa60340

**Comment by danielkerr** [comment]
ok u can override the library that is being loaded by altering the route.

**Comment by mhcwebdesign** [comment]
@danielkerr : I am not sure how this will address the issue of multiple 3rd party extensions wanting to modify a library class. Yes, many library classes are singletons, and can be replaced via the registry with something else, e.g. an extended library class with overridden methods. But this still is insufficient. 

This was also discussed here at https://github.com/opencart/opencart/discussions/11624

**Comment by danielkerr** [comment]
so u want me to make it so u can override the arguments and output

**Comment by mhcwebdesign** [comment]
Ideally the library classes should be extendable so that methods can be overridden, and then the extended class instance can be re-assigned to the registry or config. And it therefore helps if some of the private properties, such as the $data array in the system/library/cart/cart.php, were to be declared as **protected** instead of **private**.

Of course, this leaves us with the issue where different extensions want to modify the same library class. In OpenCart there were solutions for this, such as using OCmod/VQmod, or our Override Engine.

Perhaps looking at implementing more trigger points in library classes for event handlers would be a starting point. 

Or maybe the usage of a decorator design pattern could come in handy here: Multiple extensions could wrap up the same library class in decorators, that is, nested decorators, down to the original library class. See for example https://webdevetc.com/blog/php-decorator-design-pattern-and-example/  That could solve the problem where different extensions try to modify the results from a call like $this->cart->getProducts()

**Comment by gRoberts84** [comment]
I still think there is a place for OCMod/VQMod, in situations where you need to tweak part the system without totally replacing it. 

For example, I'm currently adding an additional option type to the admin, which has so far been fine with events, but when I reach needing to alter `addProduct` and `editProduct` to simply add another condition to 

`if ($product_option['type'] == 'select' || $product_option['type'] == 'radio' || $product_option['type'] == 'checkbox' || $product_option['type'] == 'image') {`

Instead of have an OCMod/VQMod statement that replaces that line, I have to override the entire function that can interfere with other extensions that may need to do the same, or vice versa. 

Generally speaking, events have proven to be great, but under certain situations, it's extremely difficult, or even impractical to use events. 

If there are easier ways to handle situations like this, I'd love to hear them, as I'm trying to push the team forward from OC 2/3 which are easier to modify.

**Comment by danielkerr** [comment]
it might be better to update the json using a script

**Comment by danielkerr** [comment]
@gRoberts84 maybe it can be fixed in the upgrade system rather then me modifying code

**Comment by danielkerr** [comment]
because u are the one overriding the classes you just extend the class u want to override. 

its not possible to extend library classes using the same method as model because they are called directly rather than using the loader class.

**Comment by mhcwebdesign** [comment]
Actually, I recently ported one of my extensions to OpenCart 4, and it had to modify the library class **system/library/cart/cart.php**, for method **getProducts**. I did this by wrapping up this class in a decorator and replaced the original instance with the decorator in the registry, all within an early controller before event handler. The decorator design pattern comes in quite handy here for cases where different extensions try to modify the same library class.

When I have some time, I'll write up an example.
